man/man1/sdsfits.1
.TH SDSFITS 1NEMO "14 January 2003"
.SH NAME
sdsfits \- convert selected plane or coordinate descriptor of SDS HDF4 to FITS
.SH SYNOPSIS
\fBsdsfits\fP [parameter=value]
.SH DESCRIPTION
\fBsdsfits\fP converts an \fIhdf(1)\fP SDS dataset into 
\fIfits(5NEMO)\fP image format. 
The current version was written to fit a particular
purpose, and will only properly convert a single 2D or 3D SDS
dataset into a FITS image/cube. Selection must be made wich SDS dataset
is converted (1=first) and optional the coordinate values can be written
instead of the SDS values. The \fIfitsbrowser\fP toolset in HDF can also
convert SDS from/to a FITS file (primary HDU).
One can also reorder the axes in the data matrix/cube, albeit at a cost.
.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBin=\fP\fIsds_file\fP
Input file (SDS HDF file)   
.TP
\fBout=\fP\fIfits_file\fP
Optional output FITS file. If no filename given, the input dataset
is merely scanned and it's rank and size are reported.
.TP
\fPselect=\fP\fIfile_number\fP
Select a \fIfile_number\fP (1 being the first) for output. Currently
only one \fIfits\fP file at a time can be extracted.
[Default: \fB1\fP].
.TP
\fPaxis=\fP\fIaxis_number\fP
If non-zero, this coordinate axis will be dumped as a 1D array 
in the fits file.
Note the first axis is the one that described the fasted running
index in the data-array (i.e. fortran array notation). A 0 is used if
the 2/3D array is written.
Default: 0.
.TP
\fPreorder=t|f\fP
Reorder the axes. Normally the data is transferred to the FITS file
"as is", without reordering, though SDS lists the axis backwards
(NZ,NY,NX: where NX is the fastest running axis in fortran sense,
and in FITS normally plotted horizontally). With \fBreorder=t\fP
all data is reordered, which can take large resources for large SDS
files.
.SH BUGS
No known bugs, but there are some limitations in the current version:
.PP
1) input SDS must be a 2D or 3D - output will be the corresponding FITS
image map/cube
.PP
2) the whole SDS is read into memory, no slicing.  Although this is fast,
it limits the size of the cube to the (virtual) memory you have.
.PP
3) Only the required FITS header keywords NAXIS, NAXIS1, NAXIS2 and NAXIS3
are written out.
.PP
.SH SEE ALSO
snaphdf(1NEMO), hdf(1), lshdf(1), fitssds(1NEMO), tsd(1NEMO), fits2sds(1HDF), phdu2sds(1HDF)
.nf

http://hdf.ncsa.uiuc.edu/fits/index.html (fitsbrowser::fits2sds,phdu2shs)
.fi
.SH HDF
The hdf format described here is the old classical "hdf4" format, not the new
"hdf5" format. Within HDF there are many "dialect", in particular the
so-called Multifile Scientific Data Set (SD) and 
Single-File Scientific Data Set (DFSD).
.SH EXAMPLES
In this example the file hdf100bg has 3 SDS datasets of 19 cells in Y, and
32 cells in X each.
.nf
% tsd hdf100bg
Found 3 scientific data sets in hdf100bg
1: R-VELOCITY AT TIME=2.00E+00(19,32) km/sec  -> [608 elements of type: 5 (FLOAT32)]
2: PHI-VELOCITY AT TIME=2.00E+00(19,32) km/sec  -> [608 elements of type: 5 (FLOAT32)]
3: DENSITY AT TIME=2.00E+00(19,32) Msolar/pc**2  -> [608 elements of type: 5 (FLOAT32)]

% sdsfits hdf100bg map1.fits 3 reorder=f
...
Data: DENSITY AT TIME=2.00E+00 Msolar/pc**2 1pe8.2 Polar
map1.fits: writing as FITS file
Data min/max = 0.00177725 362.925

% sdsfits hdf100bg map2.fits 3 reorder=t
...
Data: DENSITY AT TIME=2.00E+00 Msolar/pc**2 1pe8.2 Polar
map2.fits: writing as FITS file
reorder: 19 x 32
Data min/max = 0.00177725 362.925

.fi

.SH FILES
.SH AUTHOR
Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
19-Jan-95	V1.0 Created 	PJT
13-apr-95	V1.1 added select=, fixed bug in 2D conversion	PJT
13-apr-96	V1.2 added axis=	PJT
14-jan-03	V1.4 added reorder=	PJT
.fi


src/image/fits/sdsfits.c
/*
 *  SDSFITS: convert HDF SDS into a FITS image/cube (2d/3d only)
 *
 *      19-jan-95	V1.0  toy model, for Jim Stone		Peter Teuben
 *      13-apr-95       V1.1  also made it work for 2D          pjt
 *	13-apr-96	V1.2  optionally dump out axis descriptor pjt 
 *      10-nov-02       V1.3  add HISTORY
 */

#include <stdinc.h>
#include <getparam.h>
#include <history.h>
#include <fitsio_nemo.h>

/* #include <hdf.h> */

string defv[] = {		/* DEFAULT INPUT PARAMETERS */
    "in=???\n			Input file (SDS HDF file)",
    "out=\n			Optional output FITS file",
    "select=1\n			Select which SDS (1=first)",
    "axis=0\n                   Select this axis for 1D fits output",
    "VERSION=1.3\n		10-nov-02 PJT",
    NULL,
};

string usage="convert SDS HDF to FITS";

#define MAXRANK    3

void nemo_main()
{
  string infile, outfile, *hitem;
    char label[128], unit[128], format[128], coordsys[128];
    float *buff, *fp, fmin, fmax, *coord = 0;
    int i, j, k, nsds, ret, size, rank, dimsizes[MAXRANK];
    int nx, ny, nz, num_type, select, axis, axlen, count = 0;
    FITS *ff;

    select = getiparam("select");
    axis = getiparam("axis");

    /*
     *  Open SDS and read and display it's header
     */

    infile = getparam("in");
    nsds = DFSDndatasets(infile);
    if (nsds<0) error("Error interpreting SDS file %s",infile);
    dprintf(0,"%s: has %d scientific data sets\n",infile,nsds);


  for (count=1; count<=select; count++) {
    
    ret = DFSDgetdims(infile, &rank, dimsizes, MAXRANK);
    if (ret)
        error("Error obtaining rank and dimensions (MAXRANK=%d)",MAXRANK);
    dprintf(0,"%s: found SDS rank %d: (",infile, rank);
    for (i=rank-1, size=1; i>=0; i--) {		/* reverse order !! */
    	dprintf(0,"%d%s",dimsizes[i], i==0 ? ")" : ",");
    	size *= dimsizes[i];
    }
    dprintf(0," - total of %d \"pixels\"\n",size);
    
    for (i=rank-1, j=1; i>=0; i--, j++) {		/* reverse order !! */
    	ret = DFSDgetdimstrs(i+1,label,unit,format);
    	if (ret) {
            warning("Problem getdimstr for axis %s -- ignoring",j);
            continue;
    	}
    	dprintf(0,"Axis %d: %s %s %s\n",j,label,unit,format);
    	if (count==select && axis==j) {          /* get coordinate info */
    	    axlen = dimsizes[i];
    	    dprintf(0,"Also retrieving axis %d with length %d\n",axis,axlen);
            coord = (float *) allocate(axlen*sizeof(float));
            ret = DFSDgetdimscale(i+1,axlen,coord);    	    
    	}
    }
    ret = DFSDgetdatastrs(label,unit,format,coordsys);
    if (ret==0)
        dprintf(0,"Data: %s %s %s %s\n",label,unit,format,coordsys);
    fmax = fmin = 0.0;
  }

    /*
     * If output requested
     */

    if (hasvalue("out") && axis==0) {
        /* 
         * read all the SDS data; this must happen around here ...
         */
        ret = DFSDgetNT(&num_type);
        if (ret) error("Getting data type");
        if (num_type != 5)
            error("Can only read REAL32 floating point data now");
        buff = (float *) allocate(sizeof(float)*size);
        ret = DFSDgetdata(infile, rank, dimsizes, buff);
        if (ret) error("%d: Error reading data from SDS %s",ret,infile);

        /*
         *  revert axes for output FITS file
         */

    	outfile = getparam("out");
        dprintf(0,"%s: writing as FITS file\n",outfile);
        invert(rank,dimsizes);
        if (rank==2) {
     	    nz = 1;
	    ny = dimsizes[1];
	    nx = dimsizes[0];
        } else if (rank==3) {
	    nz = dimsizes[2];
            ny = dimsizes[1];
            nx = dimsizes[0];
        } else
            error("(%d) Cannot deal with rank != 2 or 3",rank);
	ff = fitopen(outfile,"new",rank, dimsizes);

        /*
         * and stuff it into fits format
         * but we need to revert the order of the axes
         * because of Fortran/C
         */

	fitwrhda(ff,"ORIGIN","NEMO: sdsfits w/history");

	/* only write 1 line */
	hitem = ask_history();
	fitwra(ff,"HISTORY",*hitem);

	if (fmin == fmax) {
	    fmin = fmax = buff[0];
	    for (i=0, fp=buff; i<size; i++, fp++) {
	        fmin = MIN(fmin, *fp);
	        fmax = MAX(fmax, *fp);
	    }
	    dprintf(0,"Data min/max = %g %g\n",fmin,fmax);
	}
	if (fmin==fmax) warning("DATAMIN and DATAMAX are the same");
        fitwrhdr(ff,"DATAMIN", fmin);
        fitwrhdr(ff,"DATAMAX", fmax);
	fp = buff;
	for (k=0; k<nz; k++) {
            fitsetpl(ff, 1, &k);
            for (j=0; j<ny; j++) {
                fitwrite(ff, j, fp);
                fp += nx;
            }
	}
        fitclose(ff);
    } else if (hasvalue("out") && axis > 0) {

    	outfile = getparam("out");
        dprintf(0,"%s: writing axis %d as FITS file\n",outfile,axis);
        rank = 2;
	dimsizes[0] = axlen;
        dimsizes[1] = 1;
	ff = fitopen(outfile,"new",rank, dimsizes);

        /*
         * and stuff it into fits format
         * but we need to revert the order of the axes
         * because of Fortran/C
         */

	fitwrhda(ff,"ORIGIN","NEMO: sdsfits w/ history");

	/* only write 1 line */
	hitem = ask_history();
	fitwra(ff,"HISTORY",*hitem);

	if (fmin == fmax) {
	    fmin = fmax = coord[0];
	    for (i=0; i<axlen; i++) {
	        fmin = MIN(fmin, coord[i]);
	        fmax = MAX(fmax, coord[i]);
	    }
	    dprintf(0,"Data min/max = %g %g\n",fmin,fmax);
	}
	if (fmin==fmax) warning("DATAMIN and DATAMAX are the same");
        fitwrhdr(ff,"DATAMIN", fmin);
        fitwrhdr(ff,"DATAMAX", fmax);
	fitwrite(ff, 0, coord);         /* write one line */
        fitclose(ff);
    } 

}


/* invert all elements of an array idx[n] */

invert(int n, int *idx)
{
    int i, tmp, *a, *b;

    if (n<2) return;
    
    a = &idx[0];
    b = &idx[n-1];
    for (i=0; i<n/2; i++) {
        tmp = *a;
        *a = *b;
        *b = tmp;
        a++;
        b--;
    }
}

